cloud-controller-manager/do/resources.go

```diff
@@ -34,62 +35,211 @@ import (
+	for _, droplet := range c.dropletIDMap {
+		droplet := droplet
+		droplets = append(droplets, droplet)
+	}
+
+	return droplets
+}
+
+func (c *resources) LoadBalancerByID(id string) (droplet *godo.LoadBalancer, found bool, err error) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	lb, found := c.loadBalancerIDMap[id]
+	return lb, found, nil
+}
+
+func (c *resources) LoadBalancers() []*godo.LoadBalancer {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	var lbs []*godo.LoadBalancer
+	for _, lb := range c.loadBalancerIDMap {
+		lb := lb
+		lbs = append(lbs, lb)
+	}
+
+	return lbs
+}
+
+func (c *resources) UpdateDroplets(droplets []godo.Droplet) {
+	newIDMap := make(map[int]*godo.Droplet)
+	newNameMap := make(map[string]*godo.Droplet)
+
+	for _, droplet := range droplets {
+		droplet := droplet
+		newIDMap[droplet.ID] = &droplet
+		newNameMap[droplet.Name] = &droplet
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	c.dropletIDMap = newIDMap
+	c.dropletNameMap = newNameMap
+}
+
+func (c *resources) UpdateLoadBalancers(lbs []godo.LoadBalancer) {
+	newIDMap := make(map[string]*godo.LoadBalancer)
+
+	for _, lb := range lbs {
+		lb := lb
+		newIDMap[lb.ID] = &lb
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	c.loadBalancerIDMap = newIDMap
+}
+
 // ResourcesController ensures that DO resources are properly tagged.
 type ResourcesController struct {
-	clusterIDTag string
-	kclient      kubernetes.Interface
-	gclient      *godo.Client
-	svcLister    v1lister.ServiceLister
+	clusterID string
+	kclient   kubernetes.Interface
+	gclient   *godo.Client
+	svcLister v1lister.ServiceLister
+
+	resources *resources
 }
 
-// NewResourcesController returns a new services controller responsible for
-// re-applying cluster IDs on DO load-balancers.
-func NewResourcesController(clusterIDTag string, inf v1informers.ServiceInformer,
-	k kubernetes.Interface, g *godo.Client) *ResourcesController {
+// NewResourcesController returns a new controller responsible for managing
+// DigitalOcean cloud resources.
+func NewResourcesController(
+	clusterID string,
+	r *resources,
+	inf v1informers.ServiceInformer,
+	k kubernetes.Interface,
+	g *godo.Client,
+) *ResourcesController {
 	return &ResourcesController{
-		clusterIDTag: clusterIDTag,
-		kclient:      k,
-		gclient:      g,
-		svcLister:    inf.Lister(),
+		clusterID: clusterID,
+		resources: r,
+		kclient:   k,
+		gclient:   g,
+		svcLister: inf.Lister(),
 	}
 }
 
 // Run starts the resources controller. It watches over DigitalOcean resources
 // making sure the right tags are set.
 func (r *ResourcesController) Run(stopCh <-chan struct{}) {
-	ticker := time.NewTicker(controllerSyncPeriod)
-	defer ticker.Stop()
-	// Do not wait for initial tick to pass; run immediately for reduced sync
-	// latency.
-	for tickerC := ticker.C; ; {
-		if err := r.sync(); err != nil {
-			glog.Errorf("failed to sync load-balancer tags: %s", err)
+	syncResourcesTicker := time.NewTicker(controllerSyncResourcesPeriod)
```

Comment:
The invocation of `Run()` is still [coupled to a non-empty cluster ID being provided](https://github.com/digitalocean/digitalocean-cloud-controller-manager/pull/200/files#diff-ad7d9529fb69c881ff9a4bb2a7d06ccdR111) (you may need to expand lines in Github as the line referenced by the link was unchanged) since the resource controller used to handle tags only. We need to move that check from `Initialize()` into this method.